File: src/main/java/uk/gov/justice/digital/hmpps/riskprofiler/services/EscapeDecisionTreeService.java
package uk.gov.justice.digital.hmpps.riskprofiler.services;

import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import uk.gov.justice.digital.hmpps.riskprofiler.model.Alert;
import uk.gov.justice.digital.hmpps.riskprofiler.model.EscapeProfile;
import uk.gov.justice.digital.hmpps.riskprofiler.model.RiskProfile;

import javax.validation.constraints.NotNull;
import java.util.stream.Collectors;

@Service
@Slf4j
public class EscapeDecisionTreeService {

    private final NomisService nomisService;

    public EscapeDecisionTreeService(final NomisService nomisService) {
        this.nomisService = nomisService;
    }

    public EscapeProfile getEscapeProfile(@NotNull final String nomsId) {
        final var escapeData = nomisService.getEscapeListAlertsForOffender(nomsId);

        final var splitLists =
                escapeData.stream().filter(Alert::isActive).collect(Collectors.partitioningBy(a -> a.getAlertCode().equals("XEL")));

        final var escapeListAlerts = splitLists.get(true);
        final var escapeRiskAlerts = splitLists.get(false);
        log.debug("Escape profile for {}: {} list alerts, {} risk alerts", nomsId, escapeListAlerts.size(), escapeRiskAlerts.size());
        return EscapeProfile.escapeBuilder()
                .nomsId(nomsId)
                .provisionalCategorisation(RiskProfile.DEFAULT_CAT)
                .activeEscapeList(!escapeListAlerts.isEmpty())
                .activeEscapeRisk(!escapeRiskAlerts.isEmpty())
                .escapeListAlerts(escapeListAlerts)
                .escapeRiskAlerts(escapeRiskAlerts)
                .build();
    }
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/riskprofiler/services/EscapeDecisionTreeService.kt
package uk.gov.justice.digital.hmpps.riskprofiler.services

import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import uk.gov.justice.digital.hmpps.riskprofiler.model.EscapeProfile
import uk.gov.justice.digital.hmpps.riskprofiler.model.RiskProfile
import java.util.stream.Collectors

@Service
class EscapeDecisionTreeService(private val nomisService: NomisService) {
  fun getEscapeProfile(nomsId: String): EscapeProfile {
    val escapeData = nomisService.getEscapeListAlertsForOffender(nomsId)
    val splitLists = escapeData.stream().filter { alert -> alert.active && !alert.isExpired() }
      .collect(Collectors.partitioningBy { alert -> alert.alertCode == "XEL" })
    val escapeListAlerts = splitLists[true]!!
    val escapeRiskAlerts = splitLists[false]!!
    log.debug("Escape profile for $nomsId: ${escapeListAlerts.size} list alerts, ${escapeRiskAlerts.size} risk alerts")
    return EscapeProfile(
      nomsId,
      RiskProfile.DEFAULT_CAT,
      escapeListAlerts.isNotEmpty(),
      escapeRiskAlerts.isNotEmpty(),
      escapeRiskAlerts,
      escapeListAlerts,
    )
  }

  companion object {
    private val log = LoggerFactory.getLogger(EscapeDecisionTreeService::class.java)
  }
}
